qiskit/transpiler/preset_passmanagers/level1.py

```diff
@@ -179,8 +179,7 @@ def _direction_condition(property_set):
     def _opt_control(property_set):
         return not property_set['depth_fixed_point']
 
-    if basis_gates and ('u1' in basis_gates or 'u2' in basis_gates or
-                        'u3' in basis_gates):
+    if basis_gates and 'u3' in basis_gates:
```

Comment:
I'm not sure if this change is the direction we want to be moving for the pass selection. `Optimize1qGates` only works for u1, u2, u3, u, and p, if the basis contains any 1q gates outside of that set we shouldn't use it because the results could potentially be less than optimal. This is coming up in aer where the basis is:
```
'basis_gates': [
            'u1', 'u2', 'u3', 'u', 'p', 'r', 'rx', 'ry', 'rz', 'id', 'x',
            'y', 'z', 'h', 's', 'sdg', 'sx', 't', 'tdg', 'swap', 'cx',
            'cy', 'cz', 'csx', 'cp', 'cu1', 'cu2', 'cu3', 'rxx', 'ryy',
            'rzz', 'rzx', 'ccx', 'cswap', 'mcx', 'mcy', 'mcz', 'mcsx',
            'mcphase', 'mcu1', 'mcu2', 'mcu3', 'mcrx', 'mcry', 'mcrz',
            'mcr', 'mcswap', 'unitary', 'diagonal', 'multiplexer',
            'initialize', 'kraus', 'roerror', 'delay', 'pauli', 'mcx_gray'
        ],
```
https://github.com/Qiskit/qiskit-aer/blob/master/qiskit/providers/aer/backends/qasm_simulator.py#L288-L296

The problem here (and which still existed before) is that for aer we should use `Optimize1qGatesDecomposition` because we can end up with gates post basis translation that `Optimize1qGates` can't handle.

I'm thinking there are 3 potential paths forward here we can either just hard code the optimize1qgates basis for the old ibmq basis (and some variants) and exact match. Something like:

```python
if set(basis_gates) in [{"id", "u1", "u2", "u3", "cx"}, {"id", "u2", "u3", "cx"}, {"id", "u3", "cx"},
                        {"id", "u1", "u2", "u", "cx"}, {"id", "p", "u2", "u", "cx"}, {"id", "p", "u2", "u3", "cx"},
                        {"id", "u", "cx"} {"id", "u2", "u", "cx"}]:
    _opt = [Optimize1qGates(basis_gates), CXCancellation()]
else:
    _opt = [Optimize1qGatesDecomposition(basis_gates), CXCancellation()]
```

Or alternatively we maintain a list of 1q and 2q gate names in the standard lib (similar to what we do for the random circuit function: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/circuit/random/utils.py#L60-L66 but using the name instead of the class) and look if there any 1q gates in the basis set outside of u1, u2, u3, u, and p (or gates that aren't in the std lib).

The last option is just always use `Optimize1qGatesDecomposition`. With the rest of this PR it should produce the same results as `Optimize1qGates` (for the basis that work with `Optimize1qGates`) its just a bit slower.

Reply:
> The last option is just always use Optimize1qGatesDecomposition. With the rest of this PR it should produce the same results as Optimize1qGates (for the basis that work with Optimize1qGates) its just a bit slower.

I think this is the way forward. IIRC, the main cost is in building the `Operator`, which likely has some type conversion overhead we can avoid.

Reply:
Well a bit slower is kind of an understatement. I just ran this benchmark script using current master (not with this PR applied):
```python
from qiskit.transpiler.passes import BasisTranslator
from qiskit.transpiler.passes import Optimize1qGates
from qiskit.transpiler.passes import Optimize1qGatesDecomposition
from qiskit.circuit.random import random_circuit
from qiskit.converters import circuit_to_dag
from qiskit import Aer
from qiskit.test.mock.backends import FakeManhattan
from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
import time

backend_sim = Aer.get_backend('qasm_simulator')
backend_real = FakeManhattan()

input_circuit = random_circuit(65, 4096, measure=True, seed=42)

# Aer path
print("Aer:")
aer_basis = backend_sim.configuration().basis_gates
dag = circuit_to_dag(input_circuit)
bt_aer = BasisTranslator(sel, aer_basis)
basis_dag = bt_aer.run(dag)
start = time.time()
optimize_pass = Optimize1qGates(aer_basis)
stop = time.time()
print("Create Optimize1QGates: %s" % (stop - start))
start = time.time()
result = optimize_pass.run(basis_dag)
stop = time.time()
print("Optimize 1Q Gates %s" % str(stop - start))
print("depth: %s" % result.depth())

dag = circuit_to_dag(input_circuit)
bt_aer = BasisTranslator(sel, aer_basis)
basis_dag = bt_aer.run(dag)
start = time.time()
optimize_pass = Optimize1qGatesDecomposition(aer_basis)
stop = time.time()
print("Create Optimize1qGatesDecomposition %s" % str(stop - start))
start = time.time()
result = optimize_pass.run(basis_dag)
stop = time.time()
print("Optimize 1Q decomposition %s" % str(stop - start))
print("depth: %s" % result.depth())

print("Manhattan:")
real_basis = backend_real.configuration().basis_gates
dag = circuit_to_dag(input_circuit)
bt_real = BasisTranslator(sel, real_basis)
basis_dag = bt_real.run(dag)
start = time.time()
optimize_pass = Optimize1qGates(real_basis)
stop = time.time()
print("Create Optimize1QGates: %s" % (stop - start))
start = time.time()
result = optimize_pass.run(basis_dag)
stop = time.time()
print("Optimize 1Q Gates %s" % str(stop - start))
print("depth: %s" % result.depth())

dag = circuit_to_dag(input_circuit)
bt_real = BasisTranslator(sel, real_basis)
basis_dag = bt_real.run(dag)
start = time.time()
optimize_pass = Optimize1qGatesDecomposition(real_basis)
stop = time.time()
print("Create Optimize1qGatesDecomposition %s" % str(stop - start))
start = time.time()
result = optimize_pass.run(basis_dag)
stop = time.time()
print("Optimize 1Q decomposition %s" % str(stop - start))
print("depth: %s" % result.depth())
```

which yields this result:
```
Aer Basis:
Create Optimize1QGates: 0.00016736984252929688
Optimize 1Q Gates 3.2934727668762207
depth: 10497
Create Optimize1qGatesDecomposition 0.00019860267639160156
Optimize 1Q decomposition 40.0205774307251
depth: 7746

Manhattan Basis (u1, u2, u3, id, cx):
Create Optimize1QGates: 0.00015163421630859375
Optimize 1Q Gates 19.423008680343628
depth: 46369
Create Optimize1qGatesDecomposition 0.00015783309936523438
Optimize 1Q decomposition 52.29929852485657
depth: 46369
```
so for the old ibmq basis the decompose path is ~2.7X slower. When I did profile this earlier is is spending most of it's time in constructing the Operator for the 1q gate run, IIRC a big chunk of it was actually deepcopying instructions so there is some tuning we probably can do going from a list of gate objects to an operator, but I still think it'll be slower.

I'm not opposed to this path forward it's definitely the easiest, just want us to go in with our eyes open.

Reply:
I agree the Optimize1qGatesDecomposition only solution would be the best if performance can be managed.

I also think the second option of choosing O1QD if there are 1q gates outside ['u3', 'u2', 'u1', 'u', 'p'] is not that difficult. Creating a 1q list (don't think we need a 2q list) is easy with 2 caveats,
- The list would need to be maintained
- Custom 1q gates that are used for a basis would be missed

The Aer list above up through 'tdg' would probably work.

Reply:
With help from @chriseclectic I just pushed up: https://github.com/Qiskit/qiskit-terra/pull/5682 which should speed things up quite a bit faster but still not as fast. Re-running the benchmark script above with it applied yields:

```
Aer:
Create Optimize1QGates: 0.00016021728515625
Optimize 1Q Gates 3.2395973205566406
depth: 10497
Create Optimize1qGatesDecomposition 0.0002040863037109375
Optimize 1Q decomposition 34.68856763839722
depth: 7746

Manhattan (u1, u2, u3, id, cx):
Create Optimize1QGates: 0.0001354217529296875
Optimize 1Q Gates 19.217528104782104
depth: 46369
Create Optimize1qGatesDecomposition 0.0001475811004638672
Optimize 1Q decomposition 31.421926021575928
depth: 46369
```